rust/src/database/verkle/crypto/opening.rs

```diff
@@ -0,0 +1,186 @@
+    pub fn verify_all(&self, commitment: &Commitment, values: &[Scalar]) -> bool {
+        let mut transcript = Transcript::new(b"vt");
+        let queries: Vec<VerifierQuery> = values
+            .iter()
+            .enumerate()
+            .map(|(i, value)| VerifierQuery {
+                commitment: commitment.into(),
+                point: Fr::from(i as u8),
+                result: Fr::from(value),
+            })
+            .collect();
+        self.proof
+            .check(&CRS, &PRECOMPUTED_WEIGHTS, &queries, &mut transcript)
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+
+    #[test]
+    fn open_single_can_be_used_to_proof_individual_values() {
+        let values: Vec<_> = (0..256).map(|i| Scalar::from(i + 1)).collect();
+        let commitment = Commitment::new(&values);
+
+        // Since opening is expensive, we only test a few positions.
+        for i in [0, 5, 42, 100, 254, 255] {
+            let opening = Opening::open_single(&commitment, &values, i);
+            assert!(
+                opening.verify_single(&commitment, i, &values[i as usize].clone()),
+                "failed to verify position {i}"
+            );
+            assert!(
+                !opening.verify_single(&commitment, i, &Scalar::from(i as u64 + 2)),
+                "verified wrong value at position {i}"
+            );
+        }
+    }
+
+    #[test]
+    fn using_different_values_results_in_invalid_opening() {
+        let values_1: Vec<_> = (0..256).map(|i| Scalar::from(i + 1)).collect();
+        let values_2: Vec<_> = (0..256).map(|i| Scalar::from(i + 2)).collect();
+        let commitment = Commitment::new(&values_1);
+
+        for i in [0, 5, 42, 100, 254, 255] {
+            let opening = Opening::open_single(&commitment, &values_2, i);
+            assert!(
+                !opening.verify_single(&commitment, i, &values_1[i as usize].clone()),
+                "verified value using invalid opening at position {i}"
+            );
+        }
+    }
+
+    #[test]
+    fn commitment_update_can_be_used_to_proof_modified_values() {
+        let mut values: Vec<_> = (0..256).map(|i| Scalar::from(i + 1)).collect();
+        let mut commitment = Commitment::new(&values);
+
+        for i in [0, 5, 42, 100, 254, 255] {
+            let old_value = values[i].clone();
+            values[i] = Scalar::from(i as u64 * 10);
+            commitment.update(i as u8, &old_value, &values[i]);
+            let opening = Opening::open_single(&commitment, &values, i as u8);
+
+            assert!(
+                opening.verify_single(&commitment, i as u8, &values[i].clone()),
+                "failed to verify updated position {i}"
+            );
+
+            assert!(
+                !opening.verify_single(&commitment, i as u8, &old_value),
+                "verified outdated value at position {i}"
+            );
+        }
+    }
+
+    #[test]
+    fn open_all_can_be_used_to_proof_all_values() {
```

Comment:
Should we maybe prefix all slow tests with `slow_*`? During development if you work on something different you can then run ` cargo test -- --skip slow_`.

Reply:
also fine